File: typescript/api-ts/src/scraping.ts
/**
 * Blog scraping utilities for fetching context and metadata.
 */
import * as cheerio from "cheerio";
import { MAX_CONTEXT_CHARS } from "./config.js";

/** Clean and truncate text content. */
const cleanText = (text: string) => text.replace(/\s+/g, " ").trim().slice(0, MAX_CONTEXT_CHARS);

/** Fetch and extract text content from a blog URL. */
export async function fetchBlogContext(url: string): Promise<string> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 10000);
  try {
    const response = await fetch(url, { signal: controller.signal });
    if (!response.ok) {
      throw new Error(`Request failed: ${response.status}`);
    }
    const html = await response.text();
    const $ = cheerio.load(html);
    $("script, style, noscript").remove();
    return cleanText($.text());
  } finally {
    clearTimeout(timeout);
  }
}

/** Fetch the title from a blog URL (tries og:title, twitter:title, then <title>). */
export async function fetchBlogTitle(url: string): Promise<string> {
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 10000);
  try {
    const response = await fetch(url, { signal: controller.signal });
    if (!response.ok) {
      throw new Error(`Request failed: ${response.status}`);
    }
    const html = await response.text();
    const $ = cheerio.load(html);
    const ogTitle = $('meta[property="og:title"]').attr("content");
    if (ogTitle) {
      return cleanText(ogTitle);
    }
    const twitterTitle = $('meta[name="twitter:title"]').attr("content");
    if (twitterTitle) {
      return cleanText(twitterTitle);
    }
    const title = $("title").first().text();
    return cleanText(title || "");
  } finally {
    clearTimeout(timeout);
  }
}

File: python/api-python/scraping.py
"""Blog scraping utilities for fetching context and metadata."""

import re

import requests
from bs4 import BeautifulSoup
from config import MAX_CONTEXT_CHARS


def _clean_text(text: str) -> str:
    """Clean and truncate text content."""
    text = re.sub(r"\s+", " ", text).strip()
    return text[:MAX_CONTEXT_CHARS]


def fetch_blog_context(url: str) -> str:
    """
    Fetch and extract text content from a blog URL.
    Removes scripts, styles, and other non-content elements.
    """
    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status()
    except requests.RequestException as exc:
        raise Exception(f"Failed to fetch blog URL: {exc}") from exc

    soup = BeautifulSoup(response.text, "html.parser")
    for tag in soup(["script", "style", "noscript"]):
        tag.decompose()
    text = soup.get_text(separator=" ")
    return _clean_text(text)


def fetch_blog_title(url: str) -> str:
    """
    Fetch the title from a blog URL.
    Tries og:title, twitter:title, then <title> tag.
    """
    try:
        response = requests.get(url, timeout=10)
        response.raise_for_status()
    except requests.RequestException as exc:
        raise Exception(f"Failed to fetch blog URL: {exc}") from exc

    soup = BeautifulSoup(response.text, "html.parser")

    # Try Open Graph title first
    og_title = soup.find("meta", property="og:title")
    if og_title and og_title.get("content"):
        return _clean_text(og_title["content"])

    # Try Twitter title
    twitter_title = soup.find("meta", attrs={"name": "twitter:title"})
    if twitter_title and twitter_title.get("content"):
        return _clean_text(twitter_title["content"])

    # Fall back to <title> tag
    if soup.title and soup.title.string:
        return _clean_text(soup.title.string)

    return ""
